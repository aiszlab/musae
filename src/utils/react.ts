import { Partialable } from "@aiszlab/relax/types";
import { Children, isValidElement, type ReactNode } from "react";

interface ElementProps {
  children?: ReactNode;
}

/**
 * @description
 * get the string of component children
 */
export const toReactNodeText = (children: ReactNode): Partialable<string> => {
  const _children = Children.toArray(children);

  for (const child of _children) {
    if (typeof child === "string" || typeof child === "number") {
      return child.toString();
    }

    if (isValidElement<ElementProps>(child)) {
      const textInChild = toReactNodeText(child.props.children);

      if (textInChild) {
        return textInChild;
      }
    }
  }
};
