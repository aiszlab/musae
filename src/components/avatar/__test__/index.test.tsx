import { render } from "@testing-library/react";
import { Avatar } from "..";
import React from "react";
import "@testing-library/jest-dom";
import * as relax from "@aiszlab/relax";

const { Group } = Avatar;

jest.mock("@aiszlab/relax", () => ({
  __esModule: true,
  ...jest.requireActual("@aiszlab/relax"),
}));

describe("`Avatar` Component", () => {
  test("without src", () => {
    const { container, asFragment } = render(<Avatar />);

    expect(container.querySelector(".musae-avatar")).toBeInTheDocument();
    expect(asFragment()).toMatchSnapshot();
  });

  test("with src, load error", () => {
    jest.spyOn(relax, "useImageLoader").mockReturnValue("error");

    const { container, queryByText, asFragment } = render(
      <Avatar src="https://aiszlab.com" alt="RJ" />,
    );

    expect(container.querySelector(".musae-avatar")).toBeInTheDocument();
    expect(queryByText("RJ")).toBeInTheDocument();
    expect(asFragment()).toMatchSnapshot();
  });

  test("mock image is loaded", () => {
    jest.spyOn(relax, "useImageLoader").mockReturnValue("loaded");

    const { container, queryByText, asFragment } = render(
      <Avatar src="https://aiszlab.com" alt="RJ" />,
    );

    expect(container.querySelector(".musae-avatar")).toBeInTheDocument();
    expect(queryByText("RJ")).toBeNull();
    expect(asFragment()).toMatchSnapshot();
  });
});

describe("`Avatar.Group` Component", () => {
  test("in group", () => {
    const { asFragment } = render(
      <Group>
        <Avatar alt="AA" />
        <Avatar alt="BB" />
        <Avatar alt="CC" />
      </Group>,
    );

    expect(asFragment()).toMatchSnapshot();
  });

  test("in group, but is more than limit", () => {
    const { asFragment } = render(
      <Group max={4}>
        <Avatar alt="AA" />
        <Avatar alt="BB" />
        <Avatar alt="CC" />
        <Avatar alt="DD" />
        <Avatar alt="EE" />
        <Avatar alt="FF" />
      </Group>,
    );

    expect(asFragment()).toMatchSnapshot();
  });
});
