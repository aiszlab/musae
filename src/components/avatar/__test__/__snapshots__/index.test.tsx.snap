// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`\`Avatar.Group\` Component in group 1`] = `
<DocumentFragment>
  <div
    class="musae-avatar__group group__styles.group"
  >
    <span
      class="musae-avatar theme__label.medium avatar__styles.avatar avatar__styles.medium avatar__styles.circular avatar__styles.overlapping"
      style="--backgroundColor: #EADDFF; --color: #6750A4; --borderColor: #FFF;"
    >
      <img
        alt="AA"
        class="avatar__styles.image"
        draggable="false"
      />
    </span>
    <span
      class="musae-avatar theme__label.medium avatar__styles.avatar avatar__styles.medium avatar__styles.circular avatar__styles.overlapping"
      style="--backgroundColor: #EADDFF; --color: #6750A4; --borderColor: #FFF;"
    >
      <img
        alt="BB"
        class="avatar__styles.image"
        draggable="false"
      />
    </span>
    <span
      class="musae-avatar theme__label.medium avatar__styles.avatar avatar__styles.medium avatar__styles.circular avatar__styles.overlapping"
      style="--backgroundColor: #EADDFF; --color: #6750A4; --borderColor: #FFF;"
    >
      <img
        alt="CC"
        class="avatar__styles.image"
        draggable="false"
      />
    </span>
  </div>
</DocumentFragment>
`;

exports[`\`Avatar.Group\` Component in group, but is more than limit 1`] = `
<DocumentFragment>
  <div
    class="musae-avatar__group group__styles.group"
  >
    <span
      class="musae-avatar theme__label.medium avatar__styles.avatar avatar__styles.medium avatar__styles.circular avatar__styles.overlapping"
      style="--backgroundColor: #EADDFF; --color: #6750A4; --borderColor: #FFF;"
    >
      <img
        alt="AA"
        class="avatar__styles.image"
        draggable="false"
      />
    </span>
    <span
      class="musae-avatar theme__label.medium avatar__styles.avatar avatar__styles.medium avatar__styles.circular avatar__styles.overlapping"
      style="--backgroundColor: #EADDFF; --color: #6750A4; --borderColor: #FFF;"
    >
      <img
        alt="BB"
        class="avatar__styles.image"
        draggable="false"
      />
    </span>
    <span
      class="musae-avatar theme__label.medium avatar__styles.avatar avatar__styles.medium avatar__styles.circular avatar__styles.overlapping"
      style="--backgroundColor: #EADDFF; --color: #6750A4; --borderColor: #FFF;"
    >
      <img
        alt="CC"
        class="avatar__styles.image"
        draggable="false"
      />
    </span>
    <span
      class="musae-avatar theme__label.medium avatar__styles.avatar avatar__styles.medium avatar__styles.circular avatar__styles.overlapping"
      style="--backgroundColor: #EADDFF; --color: #6750A4; --borderColor: #FFF;"
    >
      <img
        alt="DD"
        class="avatar__styles.image"
        draggable="false"
      />
    </span>
    <span
      class="musae-avatar theme__label.medium avatar__styles.avatar avatar__styles.medium avatar__styles.circular avatar__styles.overlapping"
      style="--backgroundColor: #EADDFF; --color: #6750A4; --borderColor: #FFF;"
    >
      <img
        alt="+2"
        class="avatar__styles.image"
        draggable="false"
      />
    </span>
  </div>
</DocumentFragment>
`;

exports[`\`Avatar\` Component mock image is loaded 1`] = `
<DocumentFragment>
  <span
    class="musae-avatar theme__label.medium avatar__styles.avatar avatar__styles.medium avatar__styles.circular"
    style="--backgroundColor: #EADDFF; --color: #6750A4;"
  >
    <img
      alt="RJ"
      class="avatar__styles.image"
      draggable="false"
      src="https://aiszlab.com"
    />
  </span>
</DocumentFragment>
`;

exports[`\`Avatar\` Component with src, load error 1`] = `
<DocumentFragment>
  <span
    class="musae-avatar theme__label.medium avatar__styles.avatar avatar__styles.medium avatar__styles.circular"
    style="--backgroundColor: #EADDFF; --color: #6750A4;"
  >
    RJ
  </span>
</DocumentFragment>
`;

exports[`\`Avatar\` Component without src 1`] = `
<DocumentFragment>
  <span
    class="musae-avatar theme__label.medium avatar__styles.avatar avatar__styles.medium avatar__styles.circular"
    style="--backgroundColor: #EADDFF; --color: #6750A4;"
  />
</DocumentFragment>
`;
